File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Doms/Protocol/DomsPumpControlHandler.cs
using FccDesktopAgent.Core.Adapter.Common;
using FccDesktopAgent.Core.Adapter.Doms.Jpl;

namespace FccDesktopAgent.Core.Adapter.Doms.Protocol;

/// <summary>
/// JPL message builders for pump control operations.
/// Ported from legacy ForecourtClient: EmergencyBlock(), UnblockPump(), SoftLock(), Unlock().
/// </summary>
internal static class DomsPumpControlHandler
{
    // ── JPL message names ────────────────────────────────────────────────────

    public const string EmergencyStopRequest = "FpEmergencyStop_req";
    public const string EmergencyStopResponse = "FpEmergencyStop_resp";
    public const string CancelEmergencyStopRequest = "FpCancelEmergencyStop_req";
    public const string CancelEmergencyStopResponse = "FpCancelEmergencyStop_resp";
    public const string CloseRequest = "FpClose_req";
    public const string CloseResponse = "FpClose_resp";
    public const string OpenRequest = "FpOpen_req";
    public const string OpenResponse = "FpOpen_resp";

    private const string ResultOk = "0";

    /// <summary>
    /// Build FpEmergencyStop_req. Legacy: EmergencyBlock(fpId).
    /// </summary>
    public static JplMessage BuildEmergencyStopRequest(int fpId)
        => new(Name: EmergencyStopRequest, Data: new Dictionary<string, string>
        {
            ["FpId"] = fpId.ToString("00"),
        });

    /// <summary>
    /// Build FpCancelEmergencyStop_req. Legacy: UnblockPump(fpId).
    /// </summary>
    public static JplMessage BuildCancelEmergencyStopRequest(int fpId)
        => new(Name: CancelEmergencyStopRequest, Data: new Dictionary<string, string>
        {
            ["FpId"] = fpId.ToString("00"),
        });

    /// <summary>
    /// Build FpClose_req. Legacy: SoftLock(fpId).
    /// </summary>
    public static JplMessage BuildCloseRequest(int fpId)
        => new(Name: CloseRequest, Data: new Dictionary<string, string>
        {
            ["FpId"] = fpId.ToString("00"),
        });

    /// <summary>
    /// Build FpOpen_req. Legacy: Unlock(fpId).
    /// </summary>
    public static JplMessage BuildOpenRequest(int fpId)
        => new(Name: OpenRequest, Data: new Dictionary<string, string>
        {
            ["FpId"] = fpId.ToString("00"),
        });

    /// <summary>
    /// Validate a pump control response. Returns success result.
    /// </summary>
    public static PumpControlResult ValidateControlResponse(JplMessage response)
    {
        var resultCode = response.Data?.TryGetValue("ResultCode", out var rc) == true ? rc : null;
        return resultCode == ResultOk
            ? new PumpControlResult(true)
            : new PumpControlResult(false, $"Pump control failed: ResultCode={resultCode ?? "missing"}");
    }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/doms/protocol/DomsPumpControlHandler.kt
package com.fccmiddleware.edge.adapter.doms.protocol

import com.fccmiddleware.edge.adapter.common.PumpControlResult
import com.fccmiddleware.edge.adapter.doms.jpl.JplMessage

/**
 * JPL message builders for pump control operations.
 * Ported from legacy ForecourtClient: EmergencyBlock(), UnblockPump(), SoftLock(), Unlock().
 */
object DomsPumpControlHandler {
    const val EMERGENCY_STOP_REQUEST = "FpEmergencyStop_req"
    const val EMERGENCY_STOP_RESPONSE = "FpEmergencyStop_resp"
    const val CANCEL_EMERGENCY_STOP_REQUEST = "FpCancelEmergencyStop_req"
    const val CANCEL_EMERGENCY_STOP_RESPONSE = "FpCancelEmergencyStop_resp"
    const val CLOSE_REQUEST = "FpClose_req"
    const val CLOSE_RESPONSE = "FpClose_resp"
    const val OPEN_REQUEST = "FpOpen_req"
    const val OPEN_RESPONSE = "FpOpen_resp"

    private const val RESULT_OK = "0"

    fun buildEmergencyStopRequest(fpId: Int): JplMessage =
        JplMessage(name = EMERGENCY_STOP_REQUEST, data = mapOf("FpId" to "%02d".format(fpId)))

    fun buildCancelEmergencyStopRequest(fpId: Int): JplMessage =
        JplMessage(name = CANCEL_EMERGENCY_STOP_REQUEST, data = mapOf("FpId" to "%02d".format(fpId)))

    fun buildCloseRequest(fpId: Int): JplMessage =
        JplMessage(name = CLOSE_REQUEST, data = mapOf("FpId" to "%02d".format(fpId)))

    fun buildOpenRequest(fpId: Int): JplMessage =
        JplMessage(name = OPEN_REQUEST, data = mapOf("FpId" to "%02d".format(fpId)))

    fun validateControlResponse(response: JplMessage): PumpControlResult {
        val resultCode = response.data["ResultCode"]
        return if (resultCode == RESULT_OK) {
            PumpControlResult(success = true)
        } else {
            PumpControlResult(success = false, errorMessage = "Pump control failed: ResultCode=${resultCode ?: "missing"}")
        }
    }
}
